src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -1689,11 +1682,63 @@ void EmitBackreference(RegexNode node)
                 }
                 Call(s_spanGetItemMethod);
                 LdindU2();
+                Stloc(currentCharacter);
+
                 if ((node.Options & RegexOptions.IgnoreCase) != 0)
                 {
-                    CallToLower();
+                    // if (backreferenceChar != currentChar)
+                    Ldloc(backreferenceCharacter);
+                    Ldloc(currentCharacter);
+                    Ceq();
+                    Ldc(0);
+                    Ceq();
+                    BrfalseFar(charactersMatched);
+
+                    // if (RegexCaseEquivalences.TryFindCaseEquivalencesForCharWithIBehavior(backreferenceChar, _culture, out ReadOnlySpan<char> equivalences))
```

Comment:
The _culture is going to be the same for every character that's compared, but each time TryFindCaseEquivalencesForCharWithIBehavior is going to redo all the work to figure out which mapping that culture is associated with.  Can we just store the mapping id instead and avoid that additional per-character work?

Reply:
You are right, I have now added a ref behavior parameter that acts like a cache, and will get set only if it is not yet been set, which will guarantee that we will only perform the setting of it once.

Reply:
Also, correct me if I'm wrong but passing in a ref to a the cached behavior should be safe even in multi-threading scenarios since even if two threads happen to try to evaluate the case behavior they would have both set it to the same value so I believe concurrency should not be a concern AFAIK